File: services/api/test-db-connection.js
const { Pool } = require('pg');

// Database URL from environment
const DATABASE_URL = 'postgresql://neondb_owner:npg_J2fqPGrFNHu1@ep-odd-grass-adpjlbpd-pooler.c-2.us-east-1.aws.neon.tech/neondb?sslmode=require';

async function testDatabaseConnection() {
  console.log('Testing Neon Database Connection...\n');

  const pool = new Pool({
    connectionString: DATABASE_URL,
    ssl: {
      rejectUnauthorized: false
    },
    // Connection pool settings for production
    max: 20, // Maximum number of connections
    idleTimeoutMillis: 30000,
    connectionTimeoutMillis: 5000,
  });

  try {
    // Test 1: Basic connectivity
    console.log('1. Testing basic connectivity...');
    const client = await pool.connect();
    const result = await client.query('SELECT NOW() as current_time, version() as pg_version');
    console.log('✓ Connected successfully!');
    console.log(`  Current time: ${result.rows[0].current_time}`);
    console.log(`  PostgreSQL version: ${result.rows[0].pg_version}\n`);

    // Test 2: Check for existing tables
    console.log('2. Checking existing tables...');
    const tablesResult = await client.query(`
      SELECT tablename
      FROM pg_tables
      WHERE schemaname = 'public'
      ORDER BY tablename;
    `);

    if (tablesResult.rows.length > 0) {
      console.log('✓ Found tables:');
      tablesResult.rows.forEach(row => {
        console.log(`  - ${row.tablename}`);
      });
    } else {
      console.log('⚠ No tables found in public schema');
    }
    console.log();

    // Test 3: Check for Prisma migrations table
    console.log('3. Checking Prisma migrations...');
    const migrationsResult = await client.query(`
      SELECT EXISTS (
        SELECT FROM pg_tables
        WHERE schemaname = 'public'
        AND tablename = '_prisma_migrations'
      );
    `);

    if (migrationsResult.rows[0].exists) {
      console.log('✓ Prisma migrations table exists');

      // Get migration history
      const migrationHistory = await client.query(`
        SELECT migration_name, finished_at, applied_steps_count
        FROM _prisma_migrations
        ORDER BY finished_at DESC
        LIMIT 5;
      `);

      if (migrationHistory.rows.length > 0) {
        console.log('  Recent migrations:');
        migrationHistory.rows.forEach(row => {
          console.log(`  - ${row.migration_name} (applied: ${row.finished_at})`);
        });
      }
    } else {
      console.log('⚠ Prisma migrations table not found - migrations may need to be run');
    }
    console.log();

    // Test 4: Check User and RefreshToken tables structure
    console.log('4. Checking table structures...');

    const userTableExists = await client.query(`
      SELECT EXISTS (
        SELECT FROM pg_tables
        WHERE schemaname = 'public'
        AND tablename = 'User'
      );
    `);

    if (userTableExists.rows[0].exists) {
      const userColumns = await client.query(`
        SELECT column_name, data_type, is_nullable, column_default
        FROM information_schema.columns
        WHERE table_schema = 'public' AND table_name = 'User'
        ORDER BY ordinal_position;
      `);

      console.log('✓ User table structure:');
      userColumns.rows.forEach(col => {
        console.log(`  - ${col.column_name}: ${col.data_type} (nullable: ${col.is_nullable})`);
      });
      console.log();
    } else {
      console.log('⚠ User table not found\n');
    }

    const refreshTokenTableExists = await client.query(`
      SELECT EXISTS (
        SELECT FROM pg_tables
        WHERE schemaname = 'public'
        AND tablename = 'RefreshToken'
      );
    `);

    if (refreshTokenTableExists.rows[0].exists) {
      const tokenColumns = await client.query(`
        SELECT column_name, data_type, is_nullable
        FROM information_schema.columns
        WHERE table_schema = 'public' AND table_name = 'RefreshToken'
        ORDER BY ordinal_position;
      `);

      console.log('✓ RefreshToken table structure:');
      tokenColumns.rows.forEach(col => {
        console.log(`  - ${col.column_name}: ${col.data_type} (nullable: ${col.is_nullable})`);
      });
      console.log();
    } else {
      console.log('⚠ RefreshToken table not found\n');
    }

    // Test 5: Check connection pool settings
    console.log('5. Connection pool configuration:');
    console.log(`  - Max connections: ${pool.options.max}`);
    console.log(`  - Idle timeout: ${pool.options.idleTimeoutMillis}ms`);
    console.log(`  - Connection timeout: ${pool.options.connectionTimeoutMillis}ms`);
    console.log(`  - Total clients: ${pool.totalCount}`);
    console.log(`  - Idle clients: ${pool.idleCount}`);
    console.log(`  - Waiting clients: ${pool.waitingCount}\n`);

    // Test 6: Test basic CRUD operations
    console.log('6. Testing basic CRUD operations...');

    // Create a test user
    try {
      const testUserId = 'test_' + Date.now();
      const insertResult = await client.query(
        'INSERT INTO "User" ("userId") VALUES ($1) RETURNING id, "createdAt", "userId"',
        [testUserId]
      );
      console.log('✓ INSERT successful:', insertResult.rows[0]);

      // Read the user
      const selectResult = await client.query(
        'SELECT * FROM "User" WHERE "userId" = $1',
        [testUserId]
      );
      console.log('✓ SELECT successful:', selectResult.rows[0]);

      // Update the user (if there were more fields)
      const updateResult = await client.query(
        'UPDATE "User" SET "createdAt" = NOW() WHERE "userId" = $1 RETURNING *',
        [testUserId]
      );
      console.log('✓ UPDATE successful');

      // Delete the test user
      const deleteResult = await client.query(
        'DELETE FROM "User" WHERE "userId" = $1',
        [testUserId]
      );
      console.log('✓ DELETE successful');
      console.log('✓ All CRUD operations completed successfully\n');
    } catch (crudError) {
      console.log('⚠ CRUD operations failed:', crudError.message);
      console.log('  This might indicate tables need to be created\n');
    }

    client.release();

    // Summary
    console.log('========================================');
    console.log('DATABASE CONNECTION TEST SUMMARY:');
    console.log('✓ Database is accessible');
    console.log('✓ Connection pooling is configured');

    if (!migrationsResult.rows[0].exists || !userTableExists.rows[0].exists) {
      console.log('\n⚠ REQUIRED ACTIONS:');
      console.log('  1. Run Prisma migrations to create tables:');
      console.log('     cd services/api');
      console.log('     npx prisma migrate deploy');
      console.log('  2. Generate Prisma client:');
      console.log('     npx prisma generate');
    } else {
      console.log('✓ Tables are properly set up');
    }

    console.log('========================================');

  } catch (error) {
    console.error('❌ Database connection failed:', error.message);
    console.error('Error details:', error);
    process.exit(1);
  } finally {
    await pool.end();
  }
}

// Run the test
testDatabaseConnection();

File: services/api/test-db-connection.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import psycopg2
from psycopg2 import sql
import json
from datetime import datetime
import sys

# Force UTF-8 encoding for Windows
if sys.platform == 'win32':
    sys.stdout.reconfigure(encoding='utf-8')

# Database URL
DATABASE_URL = "postgresql://neondb_owner:npg_J2fqPGrFNHu1@ep-odd-grass-adpjlbpd-pooler.c-2.us-east-1.aws.neon.tech/neondb?sslmode=require"

def test_database_connection():
    print("Testing Neon Database Connection...\n")

    try:
        # Test 1: Basic connectivity
        print("1. Testing basic connectivity...")
        conn = psycopg2.connect(DATABASE_URL)
        cursor = conn.cursor()

        cursor.execute("SELECT NOW() as current_time, version() as pg_version")
        result = cursor.fetchone()
        print("[OK] Connected successfully!")
        print(f"  Current time: {result[0]}")
        print(f"  PostgreSQL version: {result[1]}\n")

        # Test 2: Check for existing tables
        print("2. Checking existing tables...")
        cursor.execute("""
            SELECT tablename
            FROM pg_tables
            WHERE schemaname = 'public'
            ORDER BY tablename;
        """)
        tables = cursor.fetchall()

        if tables:
            print("[OK] Found tables:")
            for table in tables:
                print(f"  - {table[0]}")
        else:
            print("[WARNING] No tables found in public schema")
        print()

        # Test 3: Check for Prisma migrations table
        print("3. Checking Prisma migrations...")
        cursor.execute("""
            SELECT EXISTS (
                SELECT FROM pg_tables
                WHERE schemaname = 'public'
                AND tablename = '_prisma_migrations'
            );
        """)

        migrations_exist = cursor.fetchone()[0]
        if migrations_exist:
            print("[OK] Prisma migrations table exists")

            cursor.execute("""
                SELECT migration_name, finished_at, applied_steps_count
                FROM _prisma_migrations
                ORDER BY finished_at DESC
                LIMIT 5;
            """)
            migrations = cursor.fetchall()

            if migrations:
                print("  Recent migrations:")
                for migration in migrations:
                    print(f"  - {migration[0]} (applied: {migration[1]})")
        else:
            print("[WARNING] Prisma migrations table not found - migrations may need to be run")
        print()

        # Test 4: Check User and RefreshToken tables structure
        print("4. Checking table structures...")

        # Check User table
        cursor.execute("""
            SELECT EXISTS (
                SELECT FROM pg_tables
                WHERE schemaname = 'public'
                AND tablename = 'User'
            );
        """)

        user_table_exists = cursor.fetchone()[0]
        if user_table_exists:
            cursor.execute("""
                SELECT column_name, data_type, is_nullable, column_default
                FROM information_schema.columns
                WHERE table_schema = 'public' AND table_name = 'User'
                ORDER BY ordinal_position;
            """)
            columns = cursor.fetchall()

            print("[OK] User table structure:")
            for col in columns:
                print(f"  - {col[0]}: {col[1]} (nullable: {col[2]})")
            print()
        else:
            print("[WARNING] User table not found\n")

        # Check RefreshToken table
        cursor.execute("""
            SELECT EXISTS (
                SELECT FROM pg_tables
                WHERE schemaname = 'public'
                AND tablename = 'RefreshToken'
            );
        """)

        token_table_exists = cursor.fetchone()[0]
        if token_table_exists:
            cursor.execute("""
                SELECT column_name, data_type, is_nullable
                FROM information_schema.columns
                WHERE table_schema = 'public' AND table_name = 'RefreshToken'
                ORDER BY ordinal_position;
            """)
            columns = cursor.fetchall()

            print("[OK] RefreshToken table structure:")
            for col in columns:
                print(f"  - {col[0]}: {col[1]} (nullable: {col[2]})")
            print()
        else:
            print("[WARNING] RefreshToken table not found\n")

        # Test 5: Connection parameters
        print("5. Connection configuration:")
        print(f"  - Database: {conn.info.dbname}")
        print(f"  - User: {conn.info.user}")
        print(f"  - Host: {conn.info.host}")
        print(f"  - Port: {conn.info.port}")
        print(f"  - SSL mode: {conn.info.ssl_attribute('mode')}")
        print()

        # Test 6: Test basic CRUD operations
        if user_table_exists:
            print("6. Testing basic CRUD operations...")

            try:
                # Create a test user
                test_user_id = f"test_{datetime.now().timestamp()}"
                cursor.execute(
                    'INSERT INTO "User" ("userId") VALUES (%s) RETURNING id, "createdAt", "userId"',
                    (test_user_id,)
                )
                inserted_user = cursor.fetchone()
                print(f"[OK] INSERT successful: {inserted_user}")

                # Read the user
                cursor.execute(
                    'SELECT * FROM "User" WHERE "userId" = %s',
                    (test_user_id,)
                )
                selected_user = cursor.fetchone()
                print(f"[OK] SELECT successful: {selected_user}")

                # Update the user
                cursor.execute(
                    'UPDATE "User" SET "createdAt" = NOW() WHERE "userId" = %s RETURNING *',
                    (test_user_id,)
                )
                print("[OK] UPDATE successful")

                # Delete the test user
                cursor.execute(
                    'DELETE FROM "User" WHERE "userId" = %s',
                    (test_user_id,)
                )
                print("[OK] DELETE successful")
                print("[OK] All CRUD operations completed successfully\n")

                # Rollback since this is just a test
                conn.rollback()

            except Exception as crud_error:
                print(f"[WARNING] CRUD operations failed: {crud_error}")
                print("  This might indicate tables need to be created\n")
                conn.rollback()

        # Summary
        print("=" * 40)
        print("DATABASE CONNECTION TEST SUMMARY:")
        print("[OK] Database is accessible")
        print("[OK] Connection is properly configured")

        if not migrations_exist or not user_table_exists:
            print("\n[WARNING] REQUIRED ACTIONS:")
            print("  1. Run Prisma migrations to create tables:")
            print("     cd services/api")
            print("     npx prisma migrate deploy")
            print("  2. Generate Prisma client:")
            print("     npx prisma generate")
        else:
            print("[OK] Tables are properly set up")

        print("=" * 40)

        cursor.close()
        conn.close()

    except Exception as e:
        print(f"[ERROR] Database connection failed: {e}")
        import traceback
        traceback.print_exc()
        return False

    return True

if __name__ == "__main__":
    test_database_connection()